<!DOCTYPE html>
<html lang="en"><head>
	
	<meta name="generator" content="Hugo 0.83.1" />
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	
	<meta property="og:title" content="Django 튜토리얼 - PollApp(설문조사 앱) 6) 테스트코드 작성">
	
	
	<meta name="keywords" content="python,django"><meta property="og:title" content="Django 튜토리얼 - PollApp(설문조사 앱) 6) 테스트코드 작성" />
<meta property="og:description" content="0. 왜 테스트코드가 필요한가  계속 변경되니까 협업하니까  테스트를 만들어야하는 이유
 시간 절약 가능 : 수십개 컴포넌트 엮이면 일일히 확인 어려움 테스트는 문제를 그저 식별하는 것이 아니라 예방합니다. 테스트가 코드를 더 매력적으로 만듭니다. 테스트는 팀이 함께 일하는것을 돕습니다.  1. 쉘에서 실행 《test-driven development》
$ (myvenv) C:\IT\Django&gt;python manage.py shell Python 3.7.4 (default, Aug 9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)] on win32 Type &ldquo;help&rdquo;, &ldquo;copyright&rdquo;, &ldquo;credits&rdquo; or &ldquo;license&rdquo; for more information." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/post/20210527.html" /><meta property="article:section" content="post" />
<meta property="article:published_time" content="2021-05-27T10:28:25&#43;09:00" />
<meta property="article:modified_time" content="2021-05-27T10:28:25&#43;09:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Django 튜토리얼 - PollApp(설문조사 앱) 6) 테스트코드 작성"/>
<meta name="twitter:description" content="0. 왜 테스트코드가 필요한가  계속 변경되니까 협업하니까  테스트를 만들어야하는 이유
 시간 절약 가능 : 수십개 컴포넌트 엮이면 일일히 확인 어려움 테스트는 문제를 그저 식별하는 것이 아니라 예방합니다. 테스트가 코드를 더 매력적으로 만듭니다. 테스트는 팀이 함께 일하는것을 돕습니다.  1. 쉘에서 실행 《test-driven development》
$ (myvenv) C:\IT\Django&gt;python manage.py shell Python 3.7.4 (default, Aug 9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)] on win32 Type &ldquo;help&rdquo;, &ldquo;copyright&rdquo;, &ldquo;credits&rdquo; or &ldquo;license&rdquo; for more information."/>

	<link rel="stylesheet" type="text/css" media="screen" href="../css/normalize.css" />
	<link rel="stylesheet" type="text/css" media="screen" href="../css/main.css" />
	<link rel="stylesheet" type="text/css" media="screen" href="../css/all.css" />
	<title>Django 튜토리얼 - PollApp(설문조사 앱) 6) 테스트코드 작성 | PaperDoll - Today I Learnt</title>


</head>
<body><header>
	
	<div id="titletext"><h2 id="title"><a href="../">PaperDoll - Today I Learnt</a></h2></div>
	<div id="title-description"><div id="social">
			<nav>
				<ul></ul>
			</nav>
		</div>
	</div>
	
</header>
<main><div class="post">
	
	<div class="author">
	
	</div>
	<div class="post-header">
	
		<div class="meta">
			
			<div class="date">
				<span class="day">27</span>
				<span class="rest">May 2021</span>
			</div>
			
		</div>
		
		<div class="matter">
			<h1 class="title">Django 튜토리얼 - PollApp(설문조사 앱) 6) 테스트코드 작성</h1>
		</div>
	</div>
	<div class="markdown">
		<h2 id="0-왜-테스트코드가-필요한가">0. 왜 테스트코드가 필요한가</h2>
<ul>
<li>계속 변경되니까</li>
<li>협업하니까</li>
</ul>
<p>테스트를 만들어야하는 이유</p>
<ul>
<li>시간 절약 가능 : 수십개 컴포넌트 엮이면 일일히 확인 어려움</li>
<li>테스트는 문제를 그저 식별하는 것이 아니라 예방합니다.</li>
<li>테스트가 코드를 더 매력적으로 만듭니다.</li>
<li>테스트는 팀이 함께 일하는것을 돕습니다.</li>
</ul>
<h2 id="1-쉘에서-실행">1. 쉘에서 실행</h2>
<p>《test-driven development》</p>
<p>$ (myvenv) C:\IT\Django&gt;python manage.py shell
Python 3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)] on win32
Type &ldquo;help&rdquo;, &ldquo;copyright&rdquo;, &ldquo;credits&rdquo; or &ldquo;license&rdquo; for more information.
(InteractiveConsole)</p>
<blockquote>
<blockquote>
<blockquote>
<p>import datetime
from django.utils import timezone
from polls.models import Question</p>
</blockquote>
</blockquote>
</blockquote>
<h3 id="30일-후-날짜로-질문을-만든다면">30일 후 날짜로 질문을 만든다면</h3>
<blockquote>
<blockquote>
<blockquote>
<p>future_question = Question(pub_date = timezone.now() + datetime.timedelta(days=30))
future_question.was_published_recently()
True</p>
</blockquote>
</blockquote>
</blockquote>
<h2 id="2-위-내용을-쉘이-아닌-한줄로-실행하는-테스트-코드-만들어보기">2. 위 내용을 쉘이 아닌 한줄로 실행하는 테스트 코드 만들어보기</h2>
<ul>
<li>30일 후 pub_date를 가진 Question 인스턴스를 생성하는 메소드를 가진 django.test.TestCase 하위 클래스를 생성하고, was_published_recently()의 출력이 False가 되는지 확인하는 코드.</li>
</ul>
<p>tests.py</p>
<pre><code>import datetime

from django.test import TestCase
from django.utils import timezone

from .models import Question


class QuestionModelTests(TestCase):
    def test_was_published_recently_with_future_question(self):
        time = timezone.now() + datetime.timedelta(days=30)
        future_question = Question(pub_date=time)
    
self.assertIs(future_question.test_was_published_recently(), False)
</code></pre><h3 id="상황을-코드로-쳐서-통과되는지-보는-것임---기능을-명확하게--최대한-작성해줘야-한다">상황을 코드로 쳐서 통과되는지 보는 것임 - 기능을 명확하게 : 최대한 작성해줘야 한다</h3>
<h2 id="heading">(myvenv) C:\IT\Django&gt;py manage.py test polls
Creating test database for alias &lsquo;default&rsquo;&hellip;
System check identified no issues (0 silenced).
.</h2>
<p>Ran 1 test in 0.001s</p>
<p>OK
Destroying test database for alias &lsquo;default&rsquo;&hellip;</p>
<h3 id="뷰-테스트">뷰 테스트</h3>
<blockquote>
<blockquote>
<blockquote>
<p>response = client.get(reverse(&lsquo;polls:index&rsquo;))
Invalid HTTP_HOST header: &lsquo;testserver&rsquo;. You may need to add &lsquo;testserver&rsquo; to ALLOWED_HOSTS.
Bad Request: /polls/</p>
</blockquote>
</blockquote>
</blockquote>
<p>오류 나는 경우</p>
<p>Edit the following line in your settings.py file:</p>
<p>ALLOWED_HOSTS = [&lsquo;127.0.0.1&rsquo;, &lsquo;localhost&rsquo;]
Restart your server afterwards</p>
<p>(myvenv) (base) C:\IT\Django&gt; python manage.py shell</p>
<pre><code>Python 3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)] on win32
Type &quot;help&quot;, &quot;copyright&quot;, &quot;credits&quot; or &quot;license&quot; for more information.
(InteractiveConsole)
&gt;&gt;&gt; import datetime
&gt;&gt;&gt; from django.utils import timezone
&gt;&gt;&gt; from django.test.utils import setup_test_environment
&gt;&gt;&gt; setup_test_environment()
&gt;&gt;&gt; from django.test import Client
&gt;&gt;&gt; client = Client()
&gt;&gt;&gt; response = client.get('/')
Not Found: /
&gt;&gt;&gt; response.status_code
404
&gt;&gt;&gt; from django.urls import reverse
&gt;&gt;&gt; response = client.get(reverse('polls:index'))
Internal Server Error: /polls/
Traceback (most recent call last):
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\core\handlers\exception.py&quot;, line 47, in inner
    response = get_response(request)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\core\handlers\base.py&quot;, line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\views\generic\base.py&quot;, line 70, in view
    return self.dispatch(request, *args, **kwargs)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\views\generic\base.py&quot;, line 98, in dispatch
    return handler(request, *args, **kwargs)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\views\generic\list.py&quot;, line 142, in get
    self.object_list = self.get_queryset()
  File &quot;C:\IT\Django\polls\views.py&quot;, line 84, in get_queryset
    return Question.objects.filter(pub_date__lte=timezone.now()).order_by('-pub_date')[:5]
NameError: name 'timezone' is not defined
Traceback (most recent call last):
  File &quot;&lt;console&gt;&quot;, line 1, in &lt;module&gt;
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\test\client.py&quot;, line 742, in get
    response = super().get(path, data=data, secure=secure, **extra)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\test\client.py&quot;, line 398, in get
    **extra,
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\test\client.py&quot;, line 473, in generic
    return self.request(**r)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\test\client.py&quot;, line 719, in request
    self.check_exception(response)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\test\client.py&quot;, line 580, in check_exception
    raise exc_value
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\core\handlers\exception.py&quot;, line 47, in inner
    response = get_response(request)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\core\handlers\base.py&quot;, line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\views\generic\base.py&quot;, line 70, in view
    return self.dispatch(request, *args, **kwargs)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\views\generic\base.py&quot;, line 98, in dispatch
    return handler(request, *args, **kwargs)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\views\generic\list.py&quot;, line 142, in get
    self.object_list = self.get_queryset()
  File &quot;C:\IT\Django\polls\views.py&quot;, line 84, in get_queryset
    return Question.objects.filter(pub_date__lte=timezone.now()).order_by('-pub_date')[:5]
NameError: name 'timezone' is not defined
&gt;&gt;&gt;
KeyboardInterrupt
&gt;&gt;&gt;
KeyboardInterrupt
&gt;&gt;&gt; ^Z
</code></pre><p>$ (myvenv) (base) C:\IT\Django&gt; python manage.py shell</p>
<pre><code>Python 3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)] on win32
Type &quot;help&quot;, &quot;copyright&quot;, &quot;credits&quot; or &quot;license&quot; for more information.
(InteractiveConsole)
&gt;&gt;&gt; from django.test.utils import setup_test_environment
&gt;&gt;&gt; setup_test_environment()
&gt;&gt;&gt; ^Z
</code></pre><p>now exiting InteractiveConsole&hellip;</p>
<pre><code>(myvenv) (base) C:\IT\Django&gt; python manage.py shell
Python 3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)] on win32
Type &quot;help&quot;, &quot;copyright&quot;, &quot;credits&quot; or &quot;license&quot; for more information.
(InteractiveConsole)
&gt;&gt;&gt; from django.test.utils import setup_test_environment
&gt;&gt;&gt; setup_test_environment()
&gt;&gt;&gt; from django.test import Client
&gt;&gt;&gt; client = Client()
&gt;&gt;&gt; response = client.get('/')
Not Found: /
&gt;&gt;&gt; response.status_code
404
&gt;&gt;&gt; from django.urls import reverse
&gt;&gt;&gt; response = client.get(reverse('polls:index'))
Internal Server Error: /polls/
Traceback (most recent call last):
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\core\handlers\exception.py&quot;, line 47, in inner
    response = get_response(request)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\core\handlers\base.py&quot;, line 204, in _get_response
    response = response.render()
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\response.py&quot;, line 105, in render
    self.content = self.rendered_content
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\response.py&quot;, line 83, in rendered_content
    return template.render(context, self._request)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\backends\django.py&quot;, line 61, in render
    return self.template.render(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\base.py&quot;, line 170, in render
    return self._render(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\test\utils.py&quot;, line 100, in instrumented_test_render
    return self.nodelist.render(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\base.py&quot;, line 938, in render
    bit = node.render_annotated(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\base.py&quot;, line 905, in render_annotated
    return self.render(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\defaulttags.py&quot;, line 312, in render
    return nodelist.render(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\base.py&quot;, line 938, in render
    bit = node.render_annotated(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\base.py&quot;, line 905, in render_annotated
    return self.render(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\defaulttags.py&quot;, line 211, in render
    nodelist.append(node.render_annotated(context))
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\base.py&quot;, line 905, in render_annotated
    return self.render(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\defaulttags.py&quot;, line 446, in render
    url = reverse(view_name, args=args, kwargs=kwargs, current_app=current_app)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\urls\base.py&quot;, line 86, in reverse
    return resolver._reverse_with_prefix(view, prefix, *args, **kwargs)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\urls\resolvers.py&quot;, line 694, in _reverse_with_prefix
    raise NoReverseMatch(msg)
django.urls.exceptions.NoReverseMatch: Reverse for 'polls.detail' not found. 'polls.detail' is not a valid view function or pattern name.
Traceback (most recent call last):
  File &quot;&lt;console&gt;&quot;, line 1, in &lt;module&gt;
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\test\client.py&quot;, line 742, in get
    response = super().get(path, data=data, secure=secure, **extra)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\test\client.py&quot;, line 398, in get
    **extra,
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\test\client.py&quot;, line 473, in generic
    return self.request(**r)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\test\client.py&quot;, line 719, in request
    self.check_exception(response)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\test\client.py&quot;, line 580, in check_exception
    raise exc_value
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\core\handlers\exception.py&quot;, line 47, in inner
    response = get_response(request)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\core\handlers\base.py&quot;, line 204, in _get_response
    response = response.render()
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\response.py&quot;, line 105, in render
    self.content = self.rendered_content
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\response.py&quot;, line 83, in rendered_content
    return template.render(context, self._request)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\backends\django.py&quot;, line 61, in render
    return self.template.render(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\base.py&quot;, line 170, in render
    return self._render(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\test\utils.py&quot;, line 100, in instrumented_test_render
    return self.nodelist.render(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\base.py&quot;, line 938, in render
    bit = node.render_annotated(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\base.py&quot;, line 905, in render_annotated
    return self.render(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\defaulttags.py&quot;, line 312, in render
    return nodelist.render(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\base.py&quot;, line 938, in render
    bit = node.render_annotated(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\base.py&quot;, line 905, in render_annotated
    return self.render(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\defaulttags.py&quot;, line 211, in render
    nodelist.append(node.render_annotated(context))
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\base.py&quot;, line 905, in render_annotated
    return self.render(context)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\template\defaulttags.py&quot;, line 446, in render
    url = reverse(view_name, args=args, kwargs=kwargs, current_app=current_app)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\urls\base.py&quot;, line 86, in reverse
    return resolver._reverse_with_prefix(view, prefix, *args, **kwargs)
  File &quot;C:\IT\Django\myvenv\lib\site-packages\django\urls\resolvers.py&quot;, line 694, in _reverse_with_prefix
    raise NoReverseMatch(msg)
django.urls.exceptions.NoReverseMatch: Reverse for 'polls.detail' not found. 'polls.detail' is not a valid view function or pattern name.
&gt;&gt;&gt; response = client.get(reverse('polls:index'))
&gt;&gt;&gt;  response.status_code
  File &quot;&lt;console&gt;&quot;, line 1
    response.status_code
    ^
IndentationError: unexpected indent
&gt;&gt;&gt; response.status_code
200
&gt;&gt;&gt; response.content
b'&lt;!-- \n    &lt;ul&gt;\n    \n        &lt;li&gt;&lt;a href=&quot;/polls/2/&quot;&gt;\xeb\xb0\xa5 \xeb\xad\x90 \xeb\xa8\xb9\xec\x97\x88\xec\x96\xb4? &lt;/a&gt;&lt;/li&gt;\n    \n        &lt;li&gt;&lt;a href=&quot;/polls/1/&quot;&gt;\xec\xa7\x80\xea\xb8\x88 \xeb\xaa\x87 \xec\x8b\x9c? &lt;/a&gt;&lt;/li&gt;\n    \n    &lt;/ul&gt;\n --&gt;\n\n\n\n    &lt;ul&gt;\n    \n        &lt;li&gt;&lt;a href=&quot;/polls/2/&quot;&gt;\xeb\xb0\xa5 \xeb\xad\x90 \xeb\xa8\xb9\xec\x97\x88\xec\x96\xb4? &lt;/a&gt;&lt;/li&gt;\n    \n        &lt;li&gt;&lt;a href=&quot;/polls/1/&quot;&gt;\xec\xa7\x80\xea\xb8\x88 \xeb\xaa\x87 \xec\x8b\x9c? &lt;/a&gt;&lt;/li&gt;\n    \n    &lt;/ul&gt;\n'
&gt;&gt;&gt; response.context['latest_question_list']
&lt;QuerySet [&lt;Question: 밥 뭐 먹었어?&gt;, &lt;Question: 지금 몇 시?&gt;]&gt;
&gt;&gt;&gt;
</code></pre><p><a href="https://docs.djangoproject.com/ko/3.2/intro/tutorial05/">https://docs.djangoproject.com/ko/3.2/intro/tutorial05/</a></p>

	</div>
	
	
	
	
	
	
	
	<div class="tags">
		<div class="taxosfloating_left">
			<p>Categories</p>
		</div>
		<div class="termsfloating_right">
			<p>
			
			
			
			<a href="../categories/django/"> django </a>
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			</p>
		</div>
		<div class="clearit"></div>
		
		
		
		
	
	
	
	<div class="tags">
		<div class="taxosfloating_left">
			<p>Tags</p>
		</div>
		<div class="termsfloating_right">
			<p>
			
			
			
			
			
			
			
			<a href="../tags/django/"> django </a>
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			<a href="../tags/python/"> python </a>
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			
			</p>
		</div>
		<div class="clearit"></div>
		
		
		
		
		
	</div></div>

  </main>





</body>
</html>
